### PR TITLE
Generate Konflux components for serverless operator

### PR DESCRIPTION
**Pull request by pierDipi**
- Add support for adding components nudges automatically
- Add support for FBC builder pipeline to build operator index image

Push the SO components to the hack repository since when we cut a new branch we could potentially have 2 components with the same "release version" on 2 different branches and coordinating the merge of 2 pull requests on 2 different branches is challenging and difficult.

---

Tested with:
```
FIRST_KONFLUX_BRANCH=release-1.32 CONTAINER_ENGINE=docker make generate-ci-no-clean
```
while also enabling Konflux on Eventing 1.12-1.14 branches.

This produces:
- (for example) Eventing components with correct nudges for the `serverless-bundle-release-<xyz>`
- All components for serverless-operator, including the nudge for `serveless-bundle` to the `serverless-index`

For example:

```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  annotations:
    build.appstudio.openshift.io/pipeline: '{"name":"docker-build","bundle":"latest"}'
  name: knative-eventing-apiserver-receive-adapter-release-v112
spec:
  componentName: knative-eventing-apiserver-receive-adapter
  application: serverless-operator-release-133
   
  build-nudges-ref:
    
    - "serverless-bundle-release-133"
    
  
  source:
    git:
      url: https://github.com/openshift-knative/eventing.git
      context: 
      dockerfileUrl: openshift/ci-operator/knative-images/apiserver_receive_adapter/Dockerfile
      revision: release-v1.12
```

```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  annotations:
    build.appstudio.openshift.io/pipeline: '{"name":"docker-build","bundle":"latest"}'
  name: serverless-bundle-release-133
spec:
  componentName: serverless-bundle
  application: serverless-operator-release-133
   
  source:
    git:
      url: https://github.com/openshift-knative/serverless-operator.git
      context: 
      dockerfileUrl: olm-catalog/serverless-operator/Dockerfile
      revision: release-1.33
```

### CODE REVIEW

File: .github/workflows/release-generate-ci.yaml [hunk 1]
```diff
@@ -138,6 +138,18 @@ jobs:
...
         working-directory: ./src/github.com/openshift-knative/hack/openshift/release
+      - name: Create Pull Request - openshift-knative/hack
+        if: ${{ (github.event_name == 'push' || github.event_name == 'workflow_dispatch' || github.event_name == 'schedule') && github.ref_name == 'main' }}
+        env:
+          GH_TOKEN: ${{ secrets.SERVERLESS_QE_ROBOT }}
+          GITHUB_TOKEN: ${{ secrets.SERVERLESS_QE_ROBOT }}
+        run: |
+          set -x
+          git remote add fork "https://github.com/serverless-qe/hack.git"
+          git push "https://serverless-qe:${GH_TOKEN}@github.com/serverless-qe/hack.git" sync-konflux/main:sync-konflux/main -f
```
Comment (pierDipi) [inline .github/workflows/release-generate-ci.yaml:149]: @mgencur we need a fork for the hack repo in `serverless-qe` bot

File: .github/workflows/release-discover-branches.yaml [hunk 2]
```diff
@@ -39,6 +39,10 @@ jobs:
         working-directory: ./src/github.com/openshift-knative/hack
         run: make discover-branches ARGS="--config config/"
 
+      - name: Generate Action
+        working-directory: ./src/github.com/openshift-knative/hack
+        run: make generate-action
```
Comment (creydr) [inline .github/workflows/release-discover-branches.yaml:44]: Could we rename this action (and the Makefile target) to something like "Generate GH Actions for Konflux Manifests"? (I am not good in naming, but I have to look up often, what this `generate-action` does :( )

Comment (creydr) [reply .github/workflows/release-discover-branches.yaml:44]: But not super related to this PR though

Comment (pierDipi) [reply .github/workflows/release-discover-branches.yaml:44]: `generate-konflux-action` ?

Comment (pierDipi) [reply .github/workflows/release-discover-branches.yaml:44]: it's also doing openshift/release CI updates

Comment (pierDipi) [reply .github/workflows/release-discover-branches.yaml:44]: generate-ci-action

The actual displayed name is `Generate CI config`

Comment (pierDipi) [reply .github/workflows/release-discover-branches.yaml]: Renamed to `generate-ci-action`

File: .github/workflows/release-generate-ci-template.yaml [hunk 3]
```diff
@@ -88,3 +88,16 @@ jobs:
...
+      - name: Create Pull Request - openshift-knative/hack
+        if: ${{ (github.event_name == 'push' || github.event_name == 'workflow_dispatch' || github.event_name == 'schedule') && github.ref_name == 'main' }}
+        env:
+          GH_TOKEN: ${{ secrets.SERVERLESS_QE_ROBOT }}
+          GITHUB_TOKEN: ${{ secrets.SERVERLESS_QE_ROBOT }}
+        run: |
+          set -x
+          git remote add fork "https://github.com/serverless-qe/hack.git"
+          git push "https://serverless-qe:${GH_TOKEN}@github.com/serverless-qe/hack.git" sync-konflux/main:sync-konflux/main -f
+          gh pr create --base master --head serverless-qe:sync-konflux/main --fill-verbose || true
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply .github/workflows/release-generate-ci-template.yaml:101]: Yes

File: .github/workflows/release-discover-branches.yaml [hunk 4]
```diff
@@ -39,6 +39,10 @@ jobs:
         working-directory: ./src/github.com/openshift-knative/hack
         run: make discover-branches ARGS="--config config/"
 
+      - name: Generate Action
+        working-directory: ./src/github.com/openshift-knative/hack
+        run: make generate-ci-action
```
Comment (creydr) [inline .github/workflows/release-discover-branches.yaml:44]: cmd/discover seems to call this already: https://github.com/openshift-knative/hack/blob/65af24477502773b06320352d29118e1ae1f2d17/pkg/discover/discover.go#L46-L50

:thinking:

File: .github/workflows/release-generate-ci.yaml [hunk 5]
```diff
@@ -138,6 +138,18 @@ jobs:
...
+          gh pr create --base master --head serverless-qe:sync-konflux/main --fill-verbose || true
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply .github/workflows/release-generate-ci.yaml]: Fixed both

File: .github/workflows/release-discover-branches.yaml (hunk 4 above)
Comment (pierDipi) [reply .github/workflows/release-discover-branches.yaml]: Good catch, fixed

File: pkg/action/update_action.go [hunk 6]
```diff
@@ -37,6 +39,33 @@ func UpdateAction(cfg Config) error {
...
+			soConfig.Config.Branches = map[string]prowgen.Branch{}
+		}
+		soConfig.Config.Branches[branch] = prowgen.Branch{
+			Konflux: &prowgen.Konflux{
+				Enabled: true,
+			},
+		}
+	}
+	_, s := updateAction(soConfig)
+	steps = append(steps, s...)
```
Comment (creydr) [inline pkg/action/update_action.go:67]: For me it is not really clear, why we update the actions here for SO, when we do the same (starting at line 69 again, as we have the SO config now in the config folder)

Comment (creydr) [reply pkg/action/update_action.go:67]: Ah. This is for the release-x.y branches (which are not in the SO config in config folder), correct?

File: .github/workflows/release-generate-ci.yaml (hunk 1 above)
Comment (mgencur) [reply .github/workflows/release-generate-ci.yaml:149]: Done. https://github.com/serverless-qe/hack

File: pkg/konfluxgen/konfluxgen.go [hunk 7]
```diff
@@ -122,17 +162,28 @@ func Generate(cfg Config) error {
...
+			for _, r := range fbcImages {
+				if r.MatchString(string(ib.To)) {
+					pipeline = FBCBuild
+					break
+				}
+			}
 
-			applications[appKey][dockerfileComponentKey(c.ReleaseBuildConfiguration, ib)] = DockerfileApplicationConfig{
+			applications[appKey][dockerfileComponentKey(cfg.ComponentNameFunc, c.ReleaseBuildConfiguration, ib)] = DockerfileApplicationConfig{
 				ApplicationName:           cfg.ApplicationName,
```
Comment (creydr) [inline pkg/konfluxgen/konfluxgen.go:174]: nit: maybe sanitze the appname here too, when we do it for the component name as well :shrug:

File: pkg/action/update_action.go (hunk 6 above)
Comment (pierDipi) [reply pkg/action/update_action.go:67]: Exactly

File: pkg/konfluxgen/konfluxgen.go (hunk 7 above)
Comment (pierDipi) [reply pkg/konfluxgen/konfluxgen.go]: Done

File: pkg/action/update_action.go (hunk 6 above)
Comment (pierDipi) [reply pkg/action/update_action.go]: Added a comment to explain the why

File: pkg/konfluxgen/dockerfile-component.template.yaml [hunk 8]
```diff
@@ -3,7 +3,7 @@ kind: Component
 metadata:
   annotations:
     build.appstudio.openshift.io/pipeline: '{"name":"docker-build","bundle":"latest"}'
-  name: {{ truncate ( sanitize ( print .ProjectDirectoryImageBuildStepConfiguration.To "-" .ReleaseBuildConfiguration.Metadata.Branch ) ) }}
+  name: {{ sanitize .ComponentName }}
```
Comment (creydr) [inline pkg/konfluxgen/dockerfile-component.template.yaml:6]: nit: sanitize here probably not needed, as we sanitze it before we pass it to the template. Same probably for line 8 (application name)

Comment (pierDipi) [reply pkg/konfluxgen/dockerfile-component.template.yaml:6]: Yeah sure, but it's more robust

File: pkg/action/update_action.go (hunk 6 above)
Comment (creydr) [reply pkg/action/update_action.go]: awesome. Thanks :+1:

File: pkg/prowgen/prowgen_konflux.go [hunk 9]
```diff
@@ -69,3 +101,155 @@ func GenerateKonflux(ctx context.Context, openshiftRelease Repository, configs [
...
+
+	config, err := LoadConfig("config/serverless-operator.yaml")
+	if err != nil {
+		return nil, fmt.Errorf("failed to load config for %q: %w", r.RepositoryDirectory(), err)
+	}
+
+	konfluxVersions := make(map[string]string, len(sortedBranches))
+	for i, branch := range sortedBranches {
+		// First branch we will build the operator with konflux
+		if v := os.Getenv("FIRST_KONFLUX_BRANCH"); (v != "" && v == branch) || branch == "release-1.35" {
```
Comment (mgencur) [inline pkg/prowgen/prowgen_konflux.go:120]: Does the FIRST_KONFLUX_BRANCH parameter make sense? We already know the first konflux branch will be 1.35. And if it gets delayed (for any reason) we can fix the constant in the code. I'm probably missing why it's an env variable. Thanks

Comment (pierDipi) [reply pkg/prowgen/prowgen_konflux.go:120]: It helps with debugging and testing this flow

**Review by creydr — COMMENTED** [review]
Thanks a lot for working on this :muscle: 

Took a while for me to go through this, but looks LGTM
/lgtm

Only putting it on hold in case @mgencur has something to add or you want more reviews

/hold

**Comment by mgencur** [comment]
Looks good to me. Feel free to unhold.

**Pull request merged**